coalib/bearlib/aspects/__init__.py

```diff
@@ -0,0 +1,308 @@
+
+        self.definition = cleandoc(definition)
+        self.cause = cleandoc(cause)
+        self.example = cleandoc(example)
+        self.example_language = cleandoc(example_language)
+        self.importance_reason = cleandoc(importance_reason)
+        self.fix_suggestions = cleandoc(fix_suggestions)
+
+
+@generate_consistency_check('name', 'description', 'cast_type')
+class AspectSetting:
+    """
+    This class contains the specification about a setting that should be given
+    for an aspect.
+    """
+
+    class NoDefault:
+        pass
+
+    @enforce_signature
+    def __init__(self,
+                 name='',
+                 description='',
+                 cast_type=str,
+                 example_values=(),
+                 default=NoDefault):
+        """
+        Instantiates an ``AspectSetting`` object.
+
+        :param name:
+            Name of the setting.
+        :param description:
+            A description of what the setting is about and what it does.
+        :param cast_type:
+            The type to cast the user value into. For example, ``int`` will
+            convert and setting value as an integer.
+        :param default:
+            Default value for the setting. (Optional)
+        :param example_values:
+            A tuple of example values that help the user understand what
+            kind of values this setting takes. (Optional)
+        """
+        self.name = name
+        self.description = cleandoc(description)
+        self.cast_type = cast_type
+        self.example_values = example_values
+        self.default = default
+
+
+class Aspect:
+    '''
+    This is the ``Aspect`` base class. Each aspect may have sub-aspects,
+    which are ``Aspect`` objects themselves. An aspect also contains
+    documentation, instantiated as an ``AspectDocumentation`` object.
+
+    There is a root ``Aspect`` object in which every other aspect lives.
+
+    >>> Root = Aspect()
+
+    This is a placeholder aspect through which every other aspect can be
+    accessed.
+
+    To create new aspect, simply call the ``new_subaspect`` method with
+    the subaspect's name as a param, along with its documentation (mandatory
+    for each aspect):
+
+    >>> Root.new_subaspect(
+    ...     "Redundancy",
+    ...     doc=AspectDocumentation(
+    ...         """
+    ...         This aspect describes redundancy in your source code. Those are
+    ...         usually places where you can strip away source code in order to
+    ...         make your codebase better maintainable but keeping your full
+    ...         functionality.
+    ...         """, '...', '...', '...'
+    ...     )
+    ... )
+
+    This aspect is now accesible at ``Root.Redundancy``. Simply call
+    ``new_subaspect`` to keep creating new sub-aspects:
+
+    >>> Root.Redundancy.new_subaspect(
+    ...     "Clone",
+    ...     doc=AspectDocumentation(
+    ...         "This aspect detects clones in your source code...",
```

Comment:
an aspect doesn't detect anything! We have to be careful to demostrate a good style here because people will pick that up for a *hell lot of docs* for a *hell lot of aspects*